<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Bioinformatics Training Programme - Workshop portfolio</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./assets/img/nesi_ga.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Bioinformatics Training Programme</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Training Programme Overview</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./portfolio.html" rel="" target="" aria-current="page">
 <span class="menu-text">Workshop portfolio</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#workshop-structure-style-and-expectation" id="toc-workshop-structure-style-and-expectation" class="nav-link active" data-scroll-target="#workshop-structure-style-and-expectation">Workshop structure, style and expectation</a></li>
  <li><a href="#introductory-workshops" id="toc-introductory-workshops" class="nav-link" data-scroll-target="#introductory-workshops">Introductory workshops</a>
  <ul class="collapse">
  <li><a href="#introduction-to-bash-scripting-and-hpc-job-scheduler" id="toc-introduction-to-bash-scripting-and-hpc-job-scheduler" class="nav-link" data-scroll-target="#introduction-to-bash-scripting-and-hpc-job-scheduler">Introduction to Bash Scripting and HPC Job Scheduler</a></li>
  <li><a href="#introduction-to-the-r-programming-language" id="toc-introduction-to-the-r-programming-language" class="nav-link" data-scroll-target="#introduction-to-the-r-programming-language">Introduction to the R Programming Language</a></li>
  <li><a href="#rna-seq-data-analysis" id="toc-rna-seq-data-analysis" class="nav-link" data-scroll-target="#rna-seq-data-analysis">RNA-seq Data Analysis</a></li>
  <li><a href="#genomics-data-carpentry-introduction-to-shell" id="toc-genomics-data-carpentry-introduction-to-shell" class="nav-link" data-scroll-target="#genomics-data-carpentry-introduction-to-shell">Genomics Data Carpentry (introduction to Shell)</a></li>
  </ul></li>
  <li><a href="#intermediate-workshops" id="toc-intermediate-workshops" class="nav-link" data-scroll-target="#intermediate-workshops">Intermediate workshops</a>
  <ul class="collapse">
  <li><a href="#intermediate-shell-for-bioinformatics" id="toc-intermediate-shell-for-bioinformatics" class="nav-link" data-scroll-target="#intermediate-shell-for-bioinformatics">Intermediate Shell for Bioinformatics</a></li>
  <li><a href="#intermediate-r-advance-your-skills-with-the-r-programming-language" id="toc-intermediate-r-advance-your-skills-with-the-r-programming-language" class="nav-link" data-scroll-target="#intermediate-r-advance-your-skills-with-the-r-programming-language">Intermediate R: advance your skills with the R programming language</a></li>
  <li><a href="#outlier-analysis" id="toc-outlier-analysis" class="nav-link" data-scroll-target="#outlier-analysis">Outlier Analysis</a></li>
  </ul></li>
  <li><a href="#advanced-workshops" id="toc-advanced-workshops" class="nav-link" data-scroll-target="#advanced-workshops">Advanced workshops</a>
  <ul class="collapse">
  <li><a href="#reproducible-bioinformatics-with-nextflow-and-nf-core" id="toc-reproducible-bioinformatics-with-nextflow-and-nf-core" class="nav-link" data-scroll-target="#reproducible-bioinformatics-with-nextflow-and-nf-core">Reproducible Bioinformatics with Nextflow and nf-core</a></li>
  <li><a href="#constructing-pangenome-graphs-with-pggb" id="toc-constructing-pangenome-graphs-with-pggb" class="nav-link" data-scroll-target="#constructing-pangenome-graphs-with-pggb">Constructing Pangenome Graphs with PGGB</a></li>
  <li><a href="#scaling-gene-regulatory-networks-simulations" id="toc-scaling-gene-regulatory-networks-simulations" class="nav-link" data-scroll-target="#scaling-gene-regulatory-networks-simulations">Scaling Gene Regulatory Networks Simulations</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content column-page-left" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Workshop portfolio</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="workshop-structure-style-and-expectation" class="level2">
<h2 class="anchored" data-anchor-id="workshop-structure-style-and-expectation">Workshop structure, style and expectation</h2>
<p>All workshops are run under The Carpentries Code of Conduct to promote a safe and inclusive working environment.</p>
<p>Our training workshops are interactive live-coding events. One or more Instructors (subject matter experts who have received formal teaching training) will demonstrate by writing code in real-time while attendees follow along. Instructors provide details about what they are doing and explain the decision making and thought processes that they are going through. Attendees will follow along and work through all code at the same time as the Instructor. Short exercises, either for individuals or groups, will reinforce your learning. All workshops will also have one or more Helpers: experienced members of the training community or subject matter experts who are on hand to answer questions and solve issues as they arise. Helpers ensure that no attendee is left behind and that everyone is able to complete tasks in a timely manner.</p>
<p>Our workshops run with the help of New Zealand eScience Infrastructure (NeSI). Our partnership with NeSI ensures there is no need to install software, no compatibility issues, no minimum computing requirements. All you will need to complete the workshop is a computer with an internet connection and a web browser.</p>
<p>All workshop material, which includes written notes that cover all main points, code examples, and tasks (as well as their solutions), is provided in advance and is freely available.</p>
<p>Workshops are either in person or online (zoom), and are usually one or two days (10am - 4pm).</p>
</section>
<section id="introductory-workshops" class="level2">
<h2 class="anchored" data-anchor-id="introductory-workshops">Introductory workshops</h2>
<p>Our series of introductory workshops are designed to take you from absolute beginner to self-driving learner. Each of these workshops can be done as a standalone crash course in a particular topic, and they can also be strung together to provide a strong foundation in bioinformatics principals and practices.</p>
<section id="introduction-to-bash-scripting-and-hpc-job-scheduler" class="level3">
<h3 class="anchored" data-anchor-id="introduction-to-bash-scripting-and-hpc-job-scheduler">Introduction to Bash Scripting and HPC Job Scheduler</h3>
<p>Get started with writing your own scripts for data analysis and working on an HPC (high performance computing) environment. Some of the topics covered in the workshop are:</p>
<ul>
<li><p>Designing a variant calling workflow.</p></li>
<li><p>Automating a workflow.</p></li>
<li><p>An introduction to HPC.</p></li>
<li><p>Working with job scheduler.</p></li>
</ul>
<p>View the full workshop material here: <a href="https://genomicsaotearoa.github.io/Workshop-Bash_Scripting_And_HPC_Job_Scheduler/">Introduction to Bash Scripting and HPC Scheduler</a></p>
</section>
<section id="introduction-to-the-r-programming-language" class="level3">
<h3 class="anchored" data-anchor-id="introduction-to-the-r-programming-language">Introduction to the R Programming Language</h3>
<p>Get started with R, a highly popular programming language in the fields of biology and statistics. R is world-renowned for producing high-quality, publication-ready figures and tables. Note that this workshop is a pre-requisite for the RNA-seq Data Analysis workshop. Some of the topics covered in the workshop are:</p>
<ul>
<li><p>An introduction to R and RStudio.</p></li>
<li><p>R basics: The R language, reading data into R, storing data as objects.</p></li>
<li><p>R packages.</p></li>
<li><p>Publication-quality data presentation using ggplot2.</p></li>
<li><p>Knitr: keep track of workflow and produce easy-to-follow reports of your work.</p></li>
<li><p>Where to get more help when you are ready to do more.</p></li>
</ul>
<p>We assume the learner has no prior experience with the tools covered in the workshop. However, learners are expected to have some familiarity with biological concepts.</p>
<p>View the full workshop material here: <a href="https://genomicsaotearoa.github.io/Introduction-to-R/">Introduction to the R Programming Language</a></p>
</section>
<section id="rna-seq-data-analysis" class="level3">
<h3 class="anchored" data-anchor-id="rna-seq-data-analysis">RNA-seq Data Analysis</h3>
<p>Get started with analysing RNA-seq datasets, identifying differentially expressed genes and highlighting impacted biological processes. Some of the topics covered in the workshop are:</p>
<ul>
<li><p>Quality assessment</p></li>
<li><p>Trimming and filtering</p></li>
<li><p>Mapping and read counts</p></li>
<li><p>Differential expression analysis</p></li>
<li><p>Over-representation analysis</p></li>
</ul>
<p>View the full workshop material here: <a href="https://genomicsaotearoa.github.io/RNA-seq-workshop/">RNA-seq Data Analysis</a></p>
</section>
<section id="genomics-data-carpentry-introduction-to-shell" class="level3">
<h3 class="anchored" data-anchor-id="genomics-data-carpentry-introduction-to-shell">Genomics Data Carpentry (introduction to Shell)</h3>
<p>Learn the fundamentals of working with the Command Line Interface (CLI). Shell is a program that allows you to interact with the command line. Familiarity with the shell will allow you to access remote servers, automate tasks, and use a wide range of tools that are unavailable on a Graphical User Interface (GUI)</p>
<p>During this workshop you will learn:</p>
<ul>
<li><p>The importance of the shell.</p></li>
<li><p>How to navigate files and directories.</p></li>
<li><p>How to create, view and modify files.</p></li>
<li><p>Pipes, redirection, and scripts, which will allow you to automate your workflow.</p></li>
</ul>
<p>View the full workshop material here: <a href="https://otagobioinformaticsspringschool.github.io/shell-genomics-nesi/index.html">Introduction to the Command Line for Genomics</a></p>
</section>
</section>
<section id="intermediate-workshops" class="level2">
<h2 class="anchored" data-anchor-id="intermediate-workshops">Intermediate workshops</h2>
<section id="intermediate-shell-for-bioinformatics" class="level3">
<h3 class="anchored" data-anchor-id="intermediate-shell-for-bioinformatics">Intermediate Shell for Bioinformatics</h3>
<p>Shell overview, downloading and verifying data, inspecting and manipulating text data with Unix tools, automating file-processing. This includes:</p>
<ul>
<li><p>An overview of the Shell, UNIX and Linux.</p></li>
<li><p>Downloading data from a remote source and checking data integrity.</p></li>
<li><p>Recap navigating files and directories, and commands used in routine tasks.</p></li>
<li><p>Inspecting and manipulating data, part 1 (the head, less, grep, and sed commands).</p></li>
<li><p>Inspecting and manipulating data, part 2 (using awk and bioawk to process text).</p></li>
<li><p>Automating file processing.</p></li>
<li><p>Challenges: solve example molecular biology problems using shell scripts.</p></li>
</ul>
<p>View the full workshop material here: <a href="https://genomicsaotearoa.github.io/shell-for-bioinformatics/">Intermediate Shell for Bioinformatics</a></p>
</section>
<section id="intermediate-r-advance-your-skills-with-the-r-programming-language" class="level3">
<h3 class="anchored" data-anchor-id="intermediate-r-advance-your-skills-with-the-r-programming-language">Intermediate R: advance your skills with the R programming language</h3>
<p>Advance your skills with R! You will learn to complete R tasks with fewer lines of code, scale your analyses, and write readable code. Some of the topics covered in the workshop are:</p>
<ul>
<li><p>Introduction to relational data and the join function.</p></li>
<li><p>Working with regular expressions and functions from the stringr package.</p></li>
<li><p>Writing custom functions, working with conditional statements.</p></li>
<li><p>‘Defensive programming’.</p></li>
<li><p>Iterations - for loops, and map_*() functions.</p></li>
<li><p>The importance of data structure in R.</p></li>
</ul>
<p>View the full workshop material here: <a href="https://genomicsaotearoa.github.io/Intermediate-R/">Intermediate R: advancing your skills with the R programming language</a></p>
</section>
<section id="outlier-analysis" class="level3">
<h3 class="anchored" data-anchor-id="outlier-analysis">Outlier Analysis</h3>
<p>Identify regions of the genome that are under forms of selection during a two-day workshop. During this workshop you will:</p>
<ul>
<li><p>Download example genomic data (or prepare your own).</p></li>
<li><p>Use the PCAdapt tool to identify outlier loci within a genome.</p></li>
<li><p>Use VCFtools to identify outlier SNPS in population comparisons.</p></li>
<li><p>Use Bayescan to identify outlier SNPS based on allele frequencies.</p></li>
<li><p>Relate identified SNPS to phenotypic variation.</p></li>
<li><p>Compare the results of the different methods and discuss the results.</p></li>
</ul>
<p>The focus of this workshop is on identifying signals of selection in an example genome using the outlier analysis method. Outlier analysis assumes that the majority of the genome is under neutral selection and some loci will appear as outliers relative to this background.</p>
<p>This lesson assumes learner has no prior experience with the tools covered in the workshop. However, learners are expected to have some familiarity with biological concepts, including the concept of selection. You are expected to have some familiarity with both the R programming language and with basic command line (bash).</p>
<p>View the full workshop material here: <a href="https://genomicsaotearoa.github.io/Outlier_Analysis_Workshop/">Outlier Analysis</a></p>
</section>
</section>
<section id="advanced-workshops" class="level2">
<h2 class="anchored" data-anchor-id="advanced-workshops">Advanced workshops</h2>
<section id="reproducible-bioinformatics-with-nextflow-and-nf-core" class="level3">
<h3 class="anchored" data-anchor-id="reproducible-bioinformatics-with-nextflow-and-nf-core">Reproducible Bioinformatics with Nextflow and nf-core</h3>
<p>Reproducible research is of the utmost importance. Nextflow is workflow management software that enables the writing of scalable and reproducible scientific workflows. It can integrate various software packages and environment management systems from environment modules to Docker, Singularity, and Conda. It allows for existing pipelines written in common scripting languages to be seamlessly coupled together. Nextflow simplifies the implementation and execution of workflows on cloud or high-performance computing infrastructures.</p>
<p>In this workshop you will:</p>
<ul>
<li><p>Be introduced to Nextflow and execute an example pipeline.</p></li>
<li><p>Be introduced to nf-core, an online repository of curated pipelines.</p></li>
<li><p>Learn how to configure and customise an existing nf-core pipeline.</p></li>
<li><p>Generate metrics and reports.</p></li>
</ul>
<p>View the full workshop material here: <a href="https://genomicsaotearoa.github.io/Nextflow_Workshop/">Reproducible Bioinformatics Workflows with Nextflow and nf-core</a></p>
</section>
<section id="constructing-pangenome-graphs-with-pggb" class="level3">
<h3 class="anchored" data-anchor-id="constructing-pangenome-graphs-with-pggb">Constructing Pangenome Graphs with PGGB</h3>
<p>How to construct a pangenome graph using a popular tool (PGGB), including QC, variant extraction and short-read mapping. This workshop will include:</p>
<ul>
<li><p>An introduction to pangenome graphs.</p></li>
<li><p>Setup guide for using the tools and data in the workshop.</p></li>
<li><p>Overview of the PGGB toolkit.</p></li>
<li><p>Choosing parameters to construct a graph.</p></li>
<li><p>QC</p></li>
<li><p>Extracting variant data.</p></li>
<li><p>Mapping short reads against a pangenome graph.</p></li>
</ul>
<p>View the full workshop material here: <a href="https://genomicsaotearoa.github.io/Pangenome-Graphs-Workshop/">Unlock the Power of Pangenome Graphs</a></p>
</section>
<section id="scaling-gene-regulatory-networks-simulations" class="level3">
<h3 class="anchored" data-anchor-id="scaling-gene-regulatory-networks-simulations">Scaling Gene Regulatory Networks Simulations</h3>
<p>Simulate gene regulatory networks using R and Julia. This workshop will include: - Why simulations are valuable in systems biology,</p>
<ul>
<li><p>What regulatory networks are and how they can be modelled,</p></li>
<li><p>How to use the sismonr R package to simulate a small regulatory network, and</p></li>
<li><p>Introduction to High performance computing (HPC)</p></li>
<li><p>HPC architectures</p></li>
<li><p>Batch systems and using a Slurm scheduler</p></li>
<li><p>HPC architectures</p></li>
<li><p>How to scale up simulations on a HPC through profiling and optimisation</p></li>
</ul>
<p>View the full workshop material here: <a href="https://genomicsaotearoa.github.io/Gene_Regulatory_Networks_Simulation_Workshop/">Scaling Gene Regulatory Networks Simulations</a></p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>